packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js

```diff
@@ -3847,6 +3847,28 @@ describe('ReactHooksWithNoopRenderer', () => {
       // expect(ReactNoop.getChildren()).toEqual([span('A: 2, B: 3, C: 4')]);
     });
 
+    it('mount first state', () => {
```

Comment:
@gaearon I added your tests from https://github.com/facebook/react/pull/24438, but we do not have the "React has detected a change in the order of Hooks called by App" message with the context. On cursory investigation, that doesn't seem like something we handle yet and would need more investigation to add so I'm submitting without it in case there are ideas/suggestions for how to do that.

Reply:
I remember Andrew said we should be able to pass whether it’s mount or update from BeginWork instead of checking memoizedState to choose which dispatcher to use. Then we can fix it to use Update dispatcher for updates. Would that help? Or is the nicer warning unrelated?

Reply:
I think it's unrelated. In this case the issue is that `hooksInDev` is `null` on mount, so on update we skip the more detailed error reporting with [this check](https://github.com/facebook/react/blob/7cc97ff4c7b379a3a6c385275b98e58031bc5ca3/packages/react-reconciler/src/ReactFiberHooks.new.js#L241). I think it's just defensive and I can refactor to allow it, but I wasn't sure how `hooksInDev` is intended to work. I'll dig deeper and update the PR.

Reply:
To clarify a bit more about https://github.com/facebook/react/pull/24438#issuecomment-1112575029. It shows that the bug with `hooksInDev` is because:

- It's filled _lazily_ on first Hook call. So if a mount produced no Hook calls, this is recorded as `hooksInDev = null` ("no previous render") rather than `hooksInDev = []` ("empty Hook list").
- It's checked only in _update_ dispatchers. So if a mount dispatcher is used for updates, it won't be checked. (I think you fixed this?)
- Even in update dispatchers, it seems _checked_ lazily? I.e. having an empty list on update passes the check but should error.